### DISCUSSION TITLE
How can I run a Job just once

### DISCUSSION THREAD

**Discussion by Richard-Dufour** [shinyorg/shiny#1448] [Q&A]
First and foremost, thank you @aritchie (Allan Ritchie) for coming up with Shiny. 

I read that if you set JobInfo.IsSystemJob to false, the job should run just once.  
`bool IsSystemJob = false; //defaults to true, set to false to run once OR set it inside a job to cancel further execution`

I did, but my job runs again and again and again. How can I force it to just run one time?

A little hard to follow when the documentation is a bit behind, on https://shinylib.net/client/jobs/create/ there is a mention of setting Repeat to false (property which does no exists anymore). But I totally get it, it takes time to update docs.

Thanks for any help.

Richard

**Comment by aritchie** [comment]
Glad you like Shiny!  

So yes, I removed the repeat because it was a rare/unused/complicating scenario.  What I generally recommend is still injecting the job through AddJob, BUT have a setting class where you can enable/disable via flag.

IF you truly want add/remove, you can use the IJobManager to register a job and unregister it when you don't need it.  You can also inject the job manager into your job to decide when the unregister should happen.

Let me know if that makes sense?

**Reply by aritchie** [reply]
And IsSystemJob is not what you want.  That's more of a safety switch for internal shiny stuff right now.

**Reply by Richard-Dufour** [reply]
Thanks for the reply Allan, 

So if I understand correctly, would this work? 
Seems to work but every once in a while the job is triggered/starts again

```
  async Task IJob.Run(JobInfo jobInfo, CancellationToken cancelToken)
  {     
      await this._UploadDownloadService.DoSomethingThatMayTakeSomeTime();

      // Now completed, Cancel this Job so it doesn't fire again.
      _jobManager.Cancel(jobInfo.Identifier);
  }
```

**Reply by aritchie** [reply]
You have to register the job through IJobManager.Register, not through AddJob in MauiProgram as this will re-add the job.

In MauiProgram
```csharp
builder.Services.AddJobs();

// in your app
var _jobManager = global::Shiny.Hosting.Host.GetService<IJobManager>(); // or inject
_jobManager.Register(...);
```

**Reply by Richard-Dufour** [reply]
Thanks Allan, 

That was it, I was doing: services.AddJob() then I was, again, registering the job when needed with jobManager.Register().  So double registration really.

I spun my wheels a bit trying to change that because, I just realized, services.AddJob was also doing the work of adding the needed services. So I stole some pieces so to make the needed services available: 

```
    static MauiAppBuilder RegisterShinyServices(this MauiAppBuilder builder)
    {
        var services = builder.Services;

#if !WINDOWS
        // Do not add/register the job here. We register the job when we need it. 
        //services.AddJob(typeof(MyJobInfo));

        // Instead, add the following services here (AddJob did this for us but we need to manually do it here so to Inject our needed services)
        // See Shiny.Jobs -> public static class ServiceCollectionExtensions
        services.AddConnectivity();
        services.AddBattery();
        if (!services.HasService<IJobManager>())
        {
            services.AddDefaultRepository();            
            services.AddShinyService<JobLifecycleTask>();
            services.AddShinyService<JobManager>();
        }
#endif

        // Register other services as needed
        return builder;
    }
```

Then, when needed in my code...

`jobManager.Register(job);`

All good now. 

Give me a shout if you're ever out west (Vancouver) I owe you a beer. 
Richard
